### ISSUE TITLE
Questions Regarding Reproduction of Results in Your Paper

### ISSUE THREAD

**Issue by AnderJoeSun** [bbj-lab/protoecgnet#7]
Dear ProtoECGNet Team,

I hope you are doing well.

I followed the instructions in Readme.txt, modified label_co.py, and ran it to generate the files label_cooccur_Cat1.pt, label_cooccur_Cat3.pt, and label_cooccur_Cat4.pt separately. I then ran the following command to train the three branches and perform fusion for all categories:

        # 1d joint
        fjCheckpointsDir="/data/datasets/ptb-xl-1.0.3/checkpoints"
        fjLogsDir="/data/datasets/ptb-xl-1.0.3/Logs"
        python src/main.py \
            --job_name "1D_train_joint" \
            --epochs 200 \
            --batch_size 32 \
            --lr 0.0001 \
            --checkpoint_dir "$fjCheckpointsDir" \
            --log_dir "$fjLogsDir" \
            --save_top_k 3 \
            --patience 10 \
            --resume_checkpoint False \
            --training_stage "joint" \
            --dimension "1D" \
            --backbone "resnet1d18" \
            --single_class_prototype_per_class 5 \
            --joint_prototypes_per_border 0 \
            --sampling_rate 100 \
            --label_set "1" \
            --save_weights True \
            --seed 42 \
            --num_workers 4 \
            --dropout 0.35 \
            --l2 0.00017 \
            --scheduler_type "CosineAnnealingLR" \
            --custom_groups True \
            --proto_dim 512

        # 2dl (local) joint
        fjCheckpointsDir="/data/datasets/ptb-xl-1.0.3/checkpoints"
        fjLogsDir="/data/datasets/ptb-xl-1.0.3/Logs"
        python src/main.py \
            --job_name "2Dl_train_joint" \
            --epochs 200 \
            --batch_size 32 \
            --lr 0.0001 \
            --checkpoint_dir "$fjCheckpointsDir" \
            --log_dir "$fjLogsDir" \
            --save_top_k 3 \
            --patience 10 \
            --resume_checkpoint False \
            --training_stage "joint" \
            --dimension "2D" \
            --backbone "resnet18" \
            --single_class_prototype_per_class 18 \
            --joint_prototypes_per_border 0 \
            --sampling_rate 100 \
            --label_set "3" \
            --save_weights True \
            --seed 42 \
            --num_workers 4 \
            --dropout 0.35 \
            --l2 0.00017 \
            --scheduler_type "CosineAnnealingLR" \
            --custom_groups True \
            --proto_time_len 3 \
            --proto_dim 512

        # 2dg (global) joint
        fjCheckpointsDir="/data/datasets/ptb-xl-1.0.3/checkpoints"
        fjLogsDir="/data/datasets/ptb-xl-1.0.3/Logs"
        python src/main.py \
            --job_name "2Dg_train_joint" \
            --epochs 200 \
            --batch_size 32 \
            --lr 0.0001 \
            --checkpoint_dir "$fjCheckpointsDir" \
            --log_dir "$fjLogsDir" \
            --save_top_k 3 \
            --patience 10 \
            --resume_checkpoint False \
            --training_stage "joint" \
            --dimension "2D" \
            --backbone "resnet18" \
            --single_class_prototype_per_class 3 \
            --joint_prototypes_per_border 0 \
            --sampling_rate 100 \
            --label_set "4" \
            --save_weights True \
            --seed 42 \
            --num_workers 4 \
            --dropout 0.35 \
            --l2 0.00017 \
            --scheduler_type "CosineAnnealingLR" \
            --custom_groups True \
            --proto_time_len 32 \
            --proto_dim 512 | tee 2Dg_train_joint.$(date +%Y%m%d.%H%M%S).txt

        # fusion for all
        fjCheckpointsDir="/data/datasets/ptb-xl-1.0.3/checkpoints"
        fjLogsDir="/data/datasets/ptb-xl-1.0.3/Logs"
        python src/main.py \
            --job_name "fusion_all" \
            --epochs 200 \
            --batch_size 32 \
            --lr 0.0001 \
            --checkpoint_dir "$fjCheckpointsDir" \
            --log_dir "$fjLogsDir" \
            --save_top_k 3 \
            --patience 10 \
            --resume_checkpoint False \
            --training_stage "fusion" \
            --dimension "2D" \
            --backbone "resnet18" \
            --fusion_weights1 "$fjCheckpointsDir/1D_train_joint/last.ckpt" \
            --fusion_weights3 "$fjCheckpointsDir/2Dl_train_joint/last.ckpt" \
            --fusion_weights4 "$fjCheckpointsDir/2Dg_train_joint/last.ckpt" \
            --sampling_rate 100 \
            --label_set "all" \
            --save_weights True \
            --seed 42 \
            --num_workers 4 \
            --dropout 0.35 \
            --l2 0.00017 \
            --scheduler_type "CosineAnnealingLR" \
            --custom_groups False \
            --proto_dim 512 \
            --fusion_proto_time_len3 3 \
            --fusion_proto_time_len4 32 | tee fusion_all.$(date +%Y%m%d.%H%M%S).txt

**Below are the final results from the logs of the "fusion_all" task:**
────────────────────────────
       Test metric             DataLoader 0
────────────────────────────
        test_auc            0.9055773019790649
         test_f1            0.39886221289634705
        test_fmax           0.6208227872848511
        test_loss           0.48948433995246887
────────────────────────────
Test Metrics: [{'test_loss': 0.48948433995246887, 'test_auc': 0.9055773019790649, 'test_f1': 0.39886221289634705, 'test_fmax': 0.6208227872848511}]

**Could you kindly help answer my questions listed below?**
1. Are the commands above correct?
2. Why is the "test_auc" in the final test metrics (0.9056) different from the value reported in your paper (0.9248)? Is there something wrong?
3. The "test_f1" and "test_fmax" scores in the final test metrics are relatively low. Do you have any suggestions for improving these F1 scores?

I look forward to your reply. Thank you very much in advance for your help.

Best Regards!

**Comment by sahilsethi0105** [comment]
Hello! Thank you for your interest in our work! I will address your questions below: 
1. Yes, the commands you listed look correct for the `joint` training stage. Just make sure to include the `projection` step as described in the paper, which is necessary before training the final classifier(s). After projection, you can either train separate branch-specific classifiers using the `classifier` training_stage or a single fusion classifier using the `fusion` training_stage.
2. Your reported test AUC of 0.9056 is very close to what we have seen across different runs (particularly if we do not hyperparameter tune each stage). The value in the arXiv preprint reflects one particular run, and as with any deep learning system, there is some natural variability in results due to factors like model initialization and training dynamics. Since completing peer review, the numbers in our results section have all changed slightly, and we will now be reporting weighted AUCs with confidence intervals. The exact hyperparameters used for our final peer-reviewed paper will be in the appendix (which will be posted to arXiv in the next 3-4 weeks).
3. Your reported F1 and Fmax scores look perfectly reasonable. I addressed this in my first response here (https://github.com/bbj-lab/protoecgnet/issues/5). In brief, F1 is not typically reported for this benchmark due to label imbalance and the clinical need for class-specific thresholding. Note that Fmax was originally reported in the pre-print of the benchmarking paper (https://arxiv.org/pdf/2004.13701) but removed for the peer-reviewed version (https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9190034)

I am going to close this issue for now, but feel free to reopen or open a new one if any other questions come up!

**Comment by AnderJoeSun** [comment]
Thanks a lot!

**Comment by AnderJoeSun** [comment]
Dear Sahil,

1. I am still a bit confused about the projection stage. I understand that projection is used to explain the results of future classification or fusion. However, I’m not sure why it needs to be run before classification or fusion, especially since I noticed that neither the classification nor the fusion stages seem to require any arguments related to the projection results. So I’m wondering—if I skip the projection stage, will the fusion stage still produce valid results? And if I do need to explain the results later, can I just run the projection stage afterward to generate the prototype segments?

2. Anyway, I used the following command for the 2d partial projection stage—could you please confirm if it’s correct?  

        # 2Dl_projection_with_pretrained_weights
        fjCheckpointsDir="/data/datasets/ptb-xl-1.0.3/checkpoints"
        fjLogsDir="/data/datasets/ptb-xl-1.0.3/Logs"
        python src/main.py \
            --job_name "2Dl_projection_with_pretrained_weights" \
            --epochs 200 \
            --batch_size 32 \
            --lr 0.0001 \
            --checkpoint_dir "$fjCheckpointsDir" \
            --log_dir "$fjLogsDir" \
            --save_top_k 3 \
            --patience 10 \
            --resume_checkpoint False \
            --training_stage "projection" \
            --dimension "2D" \
            --backbone "resnet18" \
            --single_class_prototype_per_class 18 \
            --joint_prototypes_per_border 0 \
            --sampling_rate 100 \
            --label_set "3" \
            --save_weights True \
            --seed 42 \
            --num_workers 4 \
            --dropout 0.35 \
            --l2 0.00017 \
            --scheduler_type "CosineAnnealingLR" \
            --custom_groups True \
            --proto_time_len 3 \
            --proto_dim 512 \
            --pretrained_weights "$fjCheckpointsDir/2Dl_train_joint/last.ckpt"

I look forward to your reply. Thank you very much in advance for your help.

Best Regards!

Joe

**Comment by sahilsethi0105** [comment]
Hi Joe, 

1. One of the key features of prototype-based networks like ours is that they base their predictions for test samples on similarity scores to real examples from the training set. We learn the prototypes through the "joint" phase, but in order to anchor them to real training ECGs, "projection" is needed. This step ensures that the explanations generated by the model faithfully reflect how its decision is made; this is not a "post-hoc" method to explain predictions—it is a way to build interpretability directly into the decision pipeline. This is a key distinction between this line of prototype-based work and others. Happy to answer any additional questions, but first I would recommend reading the original paper on this for more background (Chen et al., NIPS 2019): https://arxiv.org/abs/1806.10574
2. That command looks correct, except for the pre-trained weights. We use the pre-trained weights from the epoch with the highest validation AUC, rather than `last.ckpt`. This is not a big deal, though (will just lower your performance a little)

Let me know if you have any more questions!

**Comment by AnderJoeSun** [comment]
Dear Sahil,

Thanks a lot for your quick response!

I understand how the projection works. What I’m asking is whether the fusion stage relies on the output of the projection. If so, why is there no argument to specify the output of the projection as an input to the fusion stage command?

I look forward to your reply. Thank you very much for your wonderful support.

Best regards,

Joe

**Comment by sahilsethi0105** [comment]
Hi Joe, 

The fusion step just needs paths to the model weights. Projection saves model weights in a .pth file. You just need to specify the path to the post-projection weights for each branch, as described in the ReadMe. You can fuse non-projected models too, which would just mean the fusion classifier is being trained on similarity scores to learned prototypes that are not anchored to training ECGs. 

Best, 
Sahil Sethi

**Comment by AnderJoeSun** [comment]
**Dear Sahil,**

Got it — thanks a lot for the explanation, that really helped!

**Best regards,**

Joe
